src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11956,6 +12025,112 @@ private ExpressionSyntax ParseNewExpression()
+                            {
+                                break;
+                            }
+                            else if (!IsPossibleCollectionElement())
+                            {
+                                goto tryAgain;
+                            }
+
+                            list.Add(this.ParseCollectionElement());
+                            continue;
+                        }
+                        else if (skipBadCollectionElementTokens(ref openBracket, list, SyntaxKind.CommaToken) == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+                    }
+                }
+                else if (skipBadCollectionElementTokens(ref openBracket, list, SyntaxKind.CommaToken) == PostSkipAction.Continue)
+                {
+                    goto tryAgain;
+                }
+            }
+
+            return _syntaxFactory.CollectionCreationExpression(
+                openBracket,
+                _pool.ToListAndFree(list),
+                this.EatToken(SyntaxKind.CloseBracketToken));
+
+            PostSkipAction skipBadCollectionElementTokens(ref SyntaxToken openBracket, SeparatedSyntaxListBuilder<CollectionElementSyntax> list, SyntaxKind expected)
+            {
+                return this.SkipBadSeparatedListTokensWithExpectedKind(ref openBracket, list,
+                    p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleCollectionElement(),
+                    p => p.CurrentToken.Kind == SyntaxKind.CloseBracketToken || p.IsTerminator(),
+                    expected);
+            }
+        }
+
+        private bool IsPossibleCollectionElement()
+        {
+            if (this.IsPossibleExpression())
+                return true;
+
+            // Checking for ':' is for error recovery when someone has a dictionary element and is missing the key part.
+            if (this.CurrentToken.Kind == SyntaxKind.ColonToken)
+                return true;
+
+            // Checking for `keyword:` is for error recovery when typing a dictionary element, but the partial
+            // identifier happens to match a keyword.
+            if (SyntaxFacts.IsReservedKeyword(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken)
+                return true;
+
+            return false;
+        }
+
+        private CollectionElementSyntax ParseCollectionElement()
+        {
+            var dotDotToken = this.TryEatToken(SyntaxKind.DotDotToken);
+            if (dotDotToken != null)
+                return _syntaxFactory.SpreadElement(dotDotToken, this.ParseExpressionCore());
+
+            // Be resilient to `keyword:val` if the user hits that while typing out a full identifier.
+            ExpressionSyntax expression;
+            if (SyntaxFacts.IsReservedKeyword(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken)
```

Comment:
Are we sure we want to do this for every keyword? What about valid attribute specifiers?

Reply:
For example, `return` or `event`.

Reply:
we coudl always tweak if that was deemed a problem. but i think th echance of that happening and mattering is low.  (in other words, it's an error recovery scenario in either event), and while the other interpretation may be more correct, it hink it's unlikely enough to happen for us to invest in it.  We coudl always change this in the future if we think it's important for better recovery.